### PR TITLE
Improve example in the intro.

### LINKED ISSUE #20844: The example in the concurrency guide is incorrect
The offending example is:

```rust
use std::thread::Thread;

fn main() {
    for _ in 0..10 {
        Thread::spawn(|| {
            println!("Hello World!");
        });
    }
}
```

With the new detaching thread semantics, this is incorrect since main can exit before all the detached threads have printed.

### PR DESCRIPTION
**Pull request by steveklabnik**
Thank you @bluss for the suggested implementation.

Fixes #20844

### CODE REVIEW

File: src/doc/intro.md [hunk 1]
```diff
@@ -392,20 +392,23 @@ Here's an example of a concurrent Rust program:
...
 }
 ```
 
-This program creates ten threads, who all print `Hello, world!`. The
-`spawn` function takes one argument, a closure, indicated by the
-double bars `||`. (The `move` keyword indicates that the closure takes
-ownership of any data it uses; we'll have more on the significance of
-this shortly.) This closure is executed in a new thread created by
-`spawn`.
+This program creates ten threads, who all print `Hello, world!`. The `scoped`
```
Comment (apasel422) [inline src/doc/intro.md]: s/who/which/

File: src/doc/intro.md [hunk 2]
```diff
@@ -392,20 +392,23 @@ Here's an example of a concurrent Rust program:
 use std::thread::Thread;
 
 fn main() {
-    for _ in range(0u, 10u) {
-        Thread::spawn(move || {
+    let guards: Vec<_> = (0..10).map(|_| {
+        Thread::scoped(move || {
```
Comment (alexcrichton) [inline src/doc/intro.md]: In this case due to not closing over anything you can actually remove the `move` keyword.

File: src/doc/intro.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/intro.md]: haha, my closures are people :wink:

**Comment by steveklabnik** [comment]
Addressed both nits.

File: src/doc/intro.md [hunk 3]
```diff
@@ -392,20 +392,21 @@ Here's an example of a concurrent Rust program:
...
 fn main() {
-    for _ in range(0u, 10u) {
-        Thread::spawn(move || {
+    let guards: Vec<_> = (0..10).map(|_| {
+        Thread::scoped(|| {
             println!("Hello, world!");
-        });
-    }
+        })
+    }).collect();
```
Comment (raindev) [inline src/doc/intro.md]: Join guards are collected in this example to show explicitly how the things work, right? The code will behave in exactly the same way if we would just leave them out. Am I correct?

File: src/doc/intro.md [hunk 4]
```diff
@@ -392,20 +392,21 @@ Here's an example of a concurrent Rust program:
 use std::thread::Thread;
 
 fn main() {
-    for _ in range(0u, 10u) {
-        Thread::spawn(move || {
+    let guards: Vec<_> = (0..10).map(|_| {
```
Comment (raindev) [inline src/doc/intro.md]: With `map` this example is much nicer :blush: :+1:

File: src/doc/intro.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/intro.md]: Well, if you don't `collect`, the `map` won't do anything, as it's lazy. If you switch to a `for`, each guard drops out of scope on each iteration, so they end up executing serially anyway.

Comment (raindev) [inline src/doc/intro.md]: Thanks, got it!

**Pull request merged**